R/PipeOpTorchFTTransformerBlock.R

```diff
@@ -0,0 +1,267 @@
+#' @section nn_module:
+#' Calls [`nn_ft_transformer_block()`] when trained.
+#' @templateVar id nn_ft_transformer_block
+#' @template pipeop_torch
+#' @template pipeop_torch_example
+#' @export
+PipeOpTorchFTTransformerBlock = R6::R6Class("PipeOpTorchFTTransformerBlock",
+  inherit = PipeOpTorch,
+  lock_objects = FALSE,
+  public = list(
+    #' @description Create a new instance of this [R6][R6::R6Class] class.
+    #' @template params_pipelines
+    initialize = function(id = "nn_ft_transformer_block", param_vals = list()) {
+      param_set = ps(
+        attention_n_heads = p_int(lower = 1L, init = 8L, tags = "train"),
+        attention_dropout = p_dbl(lower = 0, upper = 1, default = 0.2, tags = "train"),
+        attention_initialization = p_fct(levels = c("kaiming", "xavier"), init = "kaiming", tags = "train"),
+        attention_normalization = p_uty(init = nn_layer_norm, tags = "train"),
+        ffn_d_hidden = p_dbl(lower = 1, tags = "train"),
+        ffn_dropout = p_dbl(lower = 0, upper = 1, tags = "train"),
+        ffn_activation = p_uty(init = nn_reglu, custom_check = check_nn_module_generator, tags = "train"),
+        ffn_normalization = p_uty(init = nn_layer_norm, custom_check = check_nn_module_generator, tags = "train"),
+        residual_dropout = p_dbl(lower = 0, upper = 1, init = 0.0, tags = "train"),
+        prenormalization = p_lgl(init = TRUE, tags = "train"),
+        is_first_layer = p_lgl(init = FALSE, tags = "train"),
+        # query_idx = p_int(special_vals = list(NULL), default = NULL, tags = "train"),
+        # query_idx = p_uty(custom_check = function(input) check_integerish(input, null.ok = TRUE), tags = "train"),
+        query_idx = p_uty(init = NULL, custom_check = function(input) check_integerish(input, null.ok = TRUE), tags = "train"),
+        attention_bias = p_lgl(init = TRUE, tags = "train"),
+        ffn_bias_first = p_lgl(init = TRUE, tags = "train"),
+        ffn_bias_second = p_lgl(init = TRUE, tags = "train")
+      )
+
+      super$initialize(
+        id = id,
+        module_generator = nn_ft_transformer_block,
+        param_vals = param_vals,
+        param_set = param_set
+      )
+    }
+  ),
+  private = list(
+    .shapes_out = function(shapes_in, param_vals, task) {
+      if (is.null(param_vals$query_idx)) {
+        return(shapes_in[1])
+      }
+
+      shapes_out = shapes_in$input
+      # to save computation, apply the last transformer block to only the CLS token
+      shapes_out[[2L]] = 1
+      return(list(shapes_out))
+    },
+    .shape_dependent_params = function(shapes_in, param_vals, task) {
+      param_vals$d_token = shapes_in$input[3]
+      return(param_vals)
+    }
+  )
+)
+mlr3pipelines::mlr_pipeops$add("nn_ft_transformer_block", PipeOpTorchFTTransformerBlock)
+
+nn_ft_multi_head_attention = nn_module(
+  "nn_ft_multi_head_attention",
+  initialize = function(d_token, n_heads, dropout, bias, initialization) {
+    if (n_heads > 1) {
```

Comment:
no need to have this if conditional as `d_token %% n_heads == 0` if `n_heads = 1`, no?